File: Java/src/com/dbc/algorithm/Array/SubarraySumEqualsK.java
package com.dbc.algorithm.Array;

import java.util.HashMap;
import java.util.Map;

public class SubarraySumEqualsK {
    public static int subarraySum(int[] nums, int k) {
        int pre = 0, count = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        for (int num : nums) {
            pre += num;
            count += map.getOrDefault(pre - k, 0);
            map.put(pre, map.getOrDefault(pre, 0) + 1);
        }

        return count;
    }
}

File: src/tmp/Python/array/subarray_sum_equals_k.py
def subarraySum(nums: list, k: int) -> int:
    count, pre = 0, 0
    mp = {}
    mp[0] = 1
    for i in range(len(nums)):
        pre += nums[i]
        count += mp.get(pre - k, 0)
        mp[pre] = mp.get(pre, 0) + 1
    return count

if __name__ == "__main__":
    print(subarraySum(nums = [1,1,1], k = 2))